### PR TITLE
formula: add `make_deduplication_links_in`

### PR DESCRIPTION
**Pull request by cho-m**
Particularly for Java dependents that commonly duplicate JARs, e.g.
* `prestodb` can be reduced from 2GB to 600MB
* `joern` can be reduced from 1.3GB to 500MB

Also can be used for PostgreSQL dependents that have same SQL files installed to support multiple `postgresql@X` formulae.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Need to write some tests and locally experiment with.

### CODE REVIEW

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -1939,6 +1939,48 @@ def time
     end
   end
 
+  # Replace duplicate files with links to reduce disk space.
+  #
+  # FIXME: Hardlinks are not fully supported so using `hardlink: true` will only
+  # reduce the bottle size or source build but will be duplicated on bottle pour.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:1945]: What makes them not fully supported, out of interest?

**Review by MikeMcQuaid — COMMENTED** [review]
Nice idea, looking good so far! Any thoughts on what would be required to do this automatically?

**Comment by cho-m** [comment]
> Any thoughts on what would be required to do this automatically?

Symlinks are a bit risky to do automatically since they can be processed in different ways (e.g. if the symlink is resolved then it can break functionality, e.g. one reason why we may switch bin symlinks to exec scripts).

Hardlinks can be safer as they don't have a different behavior on readlink/realpath, but `brew` and filesystem support are things to consider.

File: Library/Homebrew/formula.rb (hunk 1 above)
Comment (cho-m) [reply Library/Homebrew/formula.rb:1945]: My old incomplete PR #13154 is related as BSD cp duplicates hardlinks. Trying to switch this brings its own set of problems due to bugs in other commands.

An example of bottle pour behavior is `trino` which has a bottle/estimated-unpack size of <1GB but will pour to >2GB.

There also needs to be special handling when crossing filesystem boundaries (the most extreme case being someone running on a USB formatted as (ex)FAT which doesn't support hardlinks). Of the top of my head, some that can handle this are `rsync -H` and GNU `cp --preserve=links`

Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:1945]: Thanks for context!

I think an opt-in approach like this makes sense. I think ideally handling this with automatic hardlinks in future (given they are supported in default filesystems on both macOS and Linux) and accepting that you'll get duplication for non-hard links seems ideal.

**Pull request closed without merging**